geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+Credentials are cookies, TLS client certificates, or authentication headers containing a username and password.
+
+After the server has permitted the CORS "preflight" request, the client will be able to send actual cross-origin request.
+If the server doesn’t want to allow cross-origin access, it will respond with an error message to the client.
+
+## API
+This GEP proposes to add a new field `HTTPCORSFilter` to `HTTPRouteFilter`.
+If HTTPCORSFilter is set, then the gateway will generate the response of the preflight requests and send back it to the client directly.
+For the actual cross-origin request, the gateway will add CORS headers to the response before it is sent to the client.
+
+```golang
+const (
+    // HTTPRouteFilterCORS can be used to add CORS headers to an 
+    // HTTP response before it is sent to the client.
+    //
+    // Support in HTTPRouteRule: Extended
+    //
+    // Support in HTTPBackendRef: Extended
+    HTTPRouteFilterCORS HTTPRouteFilterType = "CORS"
+)
+
+type HTTPRouteFilter struct {
+    // CORS defines a schema for a filter that responds to the
+    // cross-origin request based on HTTP response header.
+    // 
+    // Support: Extended
+    //
+    // +optional
+    CORS *HTTPCORSFilter `json:"cors,omitempty"`
+}
+
+// HTTPCORSFilter defines a filter that responds to the
+// cross-origin request based on HTTP header.
+type HTTPCORSFilter struct {
+    // AllowOrigins indicates whether the response can be shared with 
+    // requested resource from the given `Origin`.
+    // 
+    // A wildcard indicates that the requests from all `Origin` are allowed.
+    //
+    // When responding to a credentialed requests, the gateway must specify 
+    // an origin in the value of the Access-Control-Allow-Origin response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["*"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: *
+    //
+    // For a simple cross-origin interaction, the gateway sets value of the header 
+    // Access-Control-Allow-Origin same as the `Origin` header provided by the client.
+    //
+    // The status code of a successful response to a "preflight" request is an OK status (i.e., 204 or 200). 
+    // For the "preflight" request, if the request `Origin` dos not match the configured allowed origins,
+    // the gateway will return a response with error status (e.g., 403). 
+    // Moreover, the gateway will omit the relevant cross-origin response headers.
+    // Then, the client will not send actual cross-origin request.
+    // 
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["https://foo.example", "http://foo.example", "https://test.example", "http://test.example"]
```

Comment:
> I would recommend using regex to specify the matched Origins. Otherwise, there is no way to configure it to match any subdomain of a given domain, or any port of a given domain.
> 
> Also, it would be better no export the `*` wildcard directly, so the user doesn't need to study what a credentialed request means. She can use the `.*` regex instead.

The type of AllowOrigins is a list of string. 
Based on CORS spec, the valid origins include wildcard "*".
```
wildcard                         = "*"
Access-Control-Allow-Origin      = origin-or-null / wildcard
```

An origin consists of three parts: the scheme, host and port.
An HTTP host can be a wildcard, meaning you can use an asterisk () in the hostname to match any subdomain within a domain.

For example:
```
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: http-route-cors
spec:
  hostnames:
  - http.route.cors.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: http-gateway
  rules:
  - backendRefs:
    - kind: Service
      name: http-route-cors
      port: 80
    matches:
    - path:
        type: PathPrefix
        value: /resource/foo
    filters:
    - cors:
      - allowOrigins:
        - http://*.foo.example
        - https://*.foo.example
        allowMethods: 
        - GET
        - PUT
        - POST
        - DELETE
        - PATCH
        - OPTIONS
        allowHeaders: 
        - DNT
        - X-CustomHeader
        - Keep-Alive
        - User-Agent
        - X-Requested-With
        - If-Modified-Since
        - Cache-Control
        - Content-Type
        - Authorization
        exposeHeaders: 
        - Content-Security-Policy
        maxAge: 1728000
      type: CORS
```

Reply:
I added some suggested wording about possible formats for this field up above, which I think should make this clearer.

If that suggestion is okay, I'd suggest also adding an example to this godoc that shows how an Origin with scheme plus wildcarded host would work (as in the example @lianglli posted just above).

Reply:
ok

Reply:
> The host part of the origin may contain the wildcard character `*`.

According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin, if the origin given in the returned Access-Control-Allow-Origin, it should match the Origin sent by the client. So if the client sends "a.com", the server can't reply with "*.com".

Reply:
I think it again, maybe you mean "if configured with '*.com', the server should return ‘a.com’ when the client sends 'a.com' and 'b.com' when the client sends 'b.com'"?

What's the advantage of this wildcard host match expression?

Reply:
To be honest, I don't know. I'm just going by trying to clarify what I understand of the language - I haven't done a lot of work with CORS though.

Reply:
> I think it again, maybe you mean "if configured with '*.com', the server should return ‘a.com’ when the client sends 'a.com' and 'b.com' when the client sends 'b.com'"?
> 
> What's the advantage of this wildcard host match expression?

![image](https://github.com/user-attachments/assets/f4f65872-1310-4a44-a106-eb96a91f5214)

The following var $http_origin is request header 'origin'.

For a simple cross-origin interaction, the gateway sets value of the header 
Access-Control-Allow-Origin same as the `Origin` header provided by the client. However, if `AllowOrigins` has the `*` wildcard, the gateway will return a response header 
Access-Control-Allow-Origin: *.

```
    if (allowOrigins has "*") {
        set $http_origin "*";
    }

    more_set_headers 'Access-Control-Allow-Origin: $http_origin';
```

Reply:
Another example in the godoc here would serve to clarify this behavior, and remove any ambiguity.